staging/src/k8s.io/client-go/util/flowcontrol/throttle.go

```diff
@@ -66,44 +87,72 @@ type realClock struct{}
 func (realClock) Now() time.Time {
 	return time.Now()
 }
+
 func (realClock) Sleep(d time.Duration) {
 	time.Sleep(d)
 }
 
+func (realClock) Since(ts time.Time) time.Duration {
+	return time.Since(ts)
+}
+
+var (
+	_ Clock              = (*realClock)(nil)
+	_ clock.PassiveClock = (*realClock)(nil)
+)
+
 // NewTokenBucketRateLimiterWithClock is identical to NewTokenBucketRateLimiter
 // but allows an injectable clock, for testing.
 func NewTokenBucketRateLimiterWithClock(qps float32, burst int, c Clock) RateLimiter {
 	limiter := rate.NewLimiter(rate.Limit(qps), burst)
-	return newTokenBucketRateLimiter(limiter, c, qps)
+	return newTokenBucketRateLimiterWithClock(limiter, c, qps)
 }
 
-func newTokenBucketRateLimiter(limiter *rate.Limiter, c Clock, qps float32) RateLimiter {
+// NewTokenBucketPassiveRateLimiterWithClock is similar to NewTokenBucketRateLimiterWithClock
+// except that it returns a PassiveRateLimiter which does not have Accept() and Wait() methods
+// and uses a PassiveClock.
+func NewTokenBucketPassiveRateLimiterWithClock(qps float32, burst int, c clock.PassiveClock) PassiveRateLimiter {
+	limiter := rate.NewLimiter(rate.Limit(qps), burst)
+	return newTokenBucketRateLimiterWithPassiveClock(limiter, c, qps)
+}
+
+func newTokenBucketRateLimiterWithClock(limiter *rate.Limiter, c Clock, qps float32) *tokenBucketRateLimiter {
 	return &tokenBucketRateLimiter{
-		limiter: limiter,
-		clock:   c,
-		qps:     qps,
+		clockIndependentRateLimiter: clockIndependentRateLimiter{limiter: limiter, qps: qps},
+		clock:                       c,
 	}
 }
 
-func (t *tokenBucketRateLimiter) TryAccept() bool {
-	return t.limiter.AllowN(t.clock.Now(), 1)
+func newTokenBucketRateLimiterWithPassiveClock(limiter *rate.Limiter, c clock.PassiveClock, qps float32) *tokenBucketPassiveRateLimiter {
+	return &tokenBucketPassiveRateLimiter{
+		clockIndependentRateLimiter: clockIndependentRateLimiter{limiter: limiter, qps: qps},
+		clock:                       c,
+	}
 }
 
-// Accept will block until a token becomes available
-func (t *tokenBucketRateLimiter) Accept() {
-	now := t.clock.Now()
-	t.clock.Sleep(t.limiter.ReserveN(now, 1).DelayFrom(now))
+func (cirl *clockIndependentRateLimiter) Stop() {
 }
 
-func (t *tokenBucketRateLimiter) Stop() {
+func (cirl *clockIndependentRateLimiter) QPS() float32 {
+	return cirl.qps
 }
 
-func (t *tokenBucketRateLimiter) QPS() float32 {
-	return t.qps
+func (cirl *clockIndependentRateLimiter) Wait(ctx context.Context) error {
+	return cirl.limiter.Wait(ctx)
 }
 
-func (t *tokenBucketRateLimiter) Wait(ctx context.Context) error {
-	return t.limiter.Wait(ctx)
+func (tbrl *tokenBucketRateLimiter) TryAccept() bool {
```

Comment:
If you eliminated clockIndependentRateLimiter and made tokenBucketRateLimiter extend tokenBucketPassiveRateLimiter then you would not have to write the definition of TryAccept twice.

Reply:
You're right. That's what I tried initially as well. The issue that comes up is that both these types now will have a `clock` element, and more over they won't be compatible. 

Meaning that `tokenBucketRateLimiter` will have a `flowcontrol.Clock` and `tokenBucketPassiveRateLimiter` will have a `clock.PassiveClock`, and these aren't compatible (due to presence of `Sleep` method in the `flowcontrol.Clock` interface). If it were to be extended in this manner, the creation of `tokenBucketRateLimiter` would look something like this:
```go
func newTokenBucketRateLimiterWithClock(limiter *rate.Limiter, c Clock, qps float32) *tokenBucketRateLimiter {
	return &tokenBucketRateLimiter{
                // realClock{} because c here won't be compatible with clock.PassiveClock
		tokenBucketPassiveRateLimiter: tokenBucketPassiveRateLimiter{limiter: limiter, qps: qps, clock: realClock{}},
		clock:                       c,
	}
}
```
If we do this then methods like `TryAccept` will end up using `realClock` for invocations of `TryAccept` through a `tokenBucketRateLimiter` type, instead of the clock that was injected.

Because of this I thought it might help to separate out these types in this manner.

> This makes *tokenBucketPassiveRateLimiter implement Wait, which you do not want.

As per this, I can change the receiver of `Wait` from `clockIndependentRateLimiter` to `tokenBucketRateLimiter` and it should be fine.

Reply:
Huh?  Doesn't your `Clock` extend `PassiveClock`, thus every `Clock` is also a `PassiveClock`?

Reply:
I'm sorry I'm a little confused. As per this discussion https://github.com/kubernetes/kubernetes/pull/104578/#discussion_r702590494, I thought it was clear as to why `Clock` couldn't extend `PassiveClock`, unless we are okay with replacing `realClock` with `clock.RealClock`?

Reply:
But this PR already adds the `Since` method to `flowcontrol.RealClock` (lines 95--97 in the current revision), so that `flowcontrol.Clock` _can_ be PassiveClock + Since.